pkg/aws/awscontrolplane/admit_awscontrolplane.go

```diff
@@ -81,11 +83,59 @@ func (a *Admitter) Admit(request *v1beta1.AdmissionRequest) ([]admission.PatchOp
 	}
 
 	var result []admission.PatchOperation
+	var numberOfAZs int
 
-	// We only need to default if attributes are not set
-	if awsControlPlaneCR.Spec.AvailabilityZones != nil && awsControlPlaneCR.Spec.InstanceType != "" {
+	// We only need to manipulate if attributes are not set or it's a create operation
+	if awsControlPlaneCR.Spec.AvailabilityZones != nil && awsControlPlaneCR.Spec.InstanceType != "" && request.Operation != "CREATE" {
 		return result, nil
 	}
+	// We need to fetch the G8sControlPlane in case AZs need to be defaulted or the awscontrolplane is just created
+	if (aws.IsHAVersion(releaseVersion) && awsControlPlaneCR.Spec.AvailabilityZones == nil) || request.Operation == "CREATE" {
+		numberOfAZs = aws.DefaultMasterReplicas
+		fetch := func() error {
+			ctx := context.Background()
+
+			// We try to fetch the G8sControlPlane CR.
+			g8sControlPlane := &infrastructurev1alpha2.G8sControlPlane{}
+			{
+				a.Log("level", "debug", "message", fmt.Sprintf("Fetching G8sControlPlane %s", awsControlPlaneCR.Name))
+				err := a.k8sClient.CtrlClient().Get(ctx,
+					types.NamespacedName{Name: awsControlPlaneCR.GetName(),
+						Namespace: awsControlPlaneCR.GetNamespace()},
+					g8sControlPlane)
+				if err != nil {
+					return microerror.Maskf(aws.NotFoundError, "failed to fetch G8sControlplane: %v", err)
+				}
+			}
+			numberOfAZs = g8sControlPlane.Spec.Replicas
+			{
+				// If the infrastructure reference is not set, we do it here
+				if request.Operation == "CREATE" && g8sControlPlane.Spec.InfrastructureRef.Name == "" {
+					a.Log("level", "debug", "message", fmt.Sprintf("Updating infrastructure reference to  %s", awsControlPlaneCR.Name))
+					infrastructureCRRef, err := reference.GetReference(infrastructurev1alpha2scheme.Scheme, awsControlPlaneCR)
+					if err != nil {
+						return microerror.Maskf(aws.ExecutionFailedError, "failed to create reference to AWSControlplane: %v", err)
+					}
+
+					// We update the reference in the CR
+					g8sControlPlane.Spec.InfrastructureRef = *infrastructureCRRef
+					err = a.k8sClient.CtrlClient().Update(ctx, g8sControlPlane)
+					if err != nil {
+						return microerror.Maskf(aws.ExecutionFailedError, "failed to update G8sControlplane: %v", err)
```

Comment:
Execution failed is an unexpected reason and should not be exposed to the outside world. See https://github.com/giantswarm/fmt/blob/master/go/errors.md#matching-errors. Further, with this kind of masking we hide the k8s error, which might be important to know about for the caller.